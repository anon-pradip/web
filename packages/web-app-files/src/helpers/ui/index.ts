export * as fileList from './filesList'
