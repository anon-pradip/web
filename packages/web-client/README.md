# web-client

The `web-client` is a standalone package that allows you to interact with the [ownCloud Infinite Scale (oCIS)](https://github.com/owncloud/ocis/) APIs via TypeScript. It provides an abstraction layer between the server and a (web-) application that converts API data into objects with helpful types and utilities. This abstraction ensures that users of the APIs don't need in-depth knowledge about them, such as required methods or returned status codes (note that the Graph layer is currently an exception to this, meaning it doesn't process responses in any way).

The supported APIs are:

- Graph (drives, sharing, user & group management)
- OCS (capabilities & url signing)
- WebDAV (file operations)

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @ownclouders/web-client

$ pnpm add @ownclouders/web-client

$ yarn add @ownclouders/web-client
```

## Usage

To utilize the `web-client`, you must instantiate it with a base URI corresponding to your oCIS deployment and an axios instance. The axios instance, or at least its headers, is being used for all requests, requiring the inclusion of all relevant headers such as authorization.

```
import axios from axios
import { client } from '@ownclouders/web-client'

const accessToken = 'some_access_token'
const baseUri = 'some_base_uri'

const axiosClient = axios.create({
	headers: { Authorization: accessToken }
})

const { graph, ocs, webdav } = client({ axiosClient, baseURI })
```

### Graph

The following example demonstrates how to retrieve all drives accessible to the user and subsequently convert them into a `SpaceResource`. Such a resource can then be used to e.g. fetch files and folders (see example down below).

```
import { buildSpace } from '@ownclouders/web-client'

const myDrives = await graph.drives.listMyDrives()

const spaceResources = myDrives.data.value.map(buildSpace)
```

### OCS

The following examples demonstrate how to fetch capabilities and sign URLs.

```
const capablities = await ocs.getCapabilities()

const signedUrl = await ocs.signUrl('some_url_to_sign', 'your_username')
```

### WebDAV

The following example demonstrates how to list all resources of a given `SpaceResource` (see above how to fetch space resources).

```
const { resource, children } = await webdav.listFiles(spaceResource)
```
