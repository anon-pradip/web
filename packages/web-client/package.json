{
  "name": "@ownclouders/web-client",
  "version": "0.0.3",
  "description": "ownCloud web client",
  "license": "AGPL-3.0",
  "private": false,
  "type": "module",
  "author": "ownCloud GmbH <devops@owncloud.com>",
  "homepage": "https://github.com/owncloud/web/tree/master/packages/web-client",
  "repository": {
    "type": "git",
    "url": "https://github.com/owncloud/web",
    "directory": "packages/web-client"
  },
  "files": [
    "dist/"
  ],
  "publishConfig": {
    "directory": "package",
    "linkDirectory": false,
    "exports": {
      ".": {
        "default": "./dist/web-client.js",
        "require": "./dist/web-client.cjs",
        "types": "./dist/src/index.d.ts"
      },
      "./graph": {
        "default": "./dist/web-client/graph.js",
        "require": "./dist/web-client/graph.cjs",
        "types": "./dist/src/graph/index.d.ts"
      },
      "./graph/generated": {
        "default": "./dist/web-client/graph/generated.js",
        "require": "./dist/web-client/graph/generated.cjs",
        "types": "./dist/src/graph/generated/index.d.ts"
      },
      "./ocs": {
        "default": "./dist/web-client/ocs.js",
        "require": "./dist/web-client/ocs.cjs",
        "types": "./dist/src/ocs/index.d.ts"
      },
      "./sse": {
        "default": "./dist/web-client/sse.js",
        "require": "./dist/web-client/sse.cjs",
        "types": "./dist/src/sse/index.d.ts"
      },
      "./webdav": {
        "default": "./dist/web-client/webdav.js",
        "require": "./dist/web-client/webdav.cjs",
        "types": "./dist/src/webdav/index.d.ts"
      }
    }
  },
  "exports": {
    ".": {
      "default": "./src/index.ts"
    },
    "./graph": {
      "default": "./src/graph/index.ts"
    },
    "./graph/generated": {
      "default": "./src/graph/generated/index.ts"
    },
    "./ocs": {
      "default": "./src/ocs/index.ts"
    },
    "./sse": {
      "default": "./src/sse/index.ts"
    },
    "./webdav": {
      "default": "./src/webdav/index.ts"
    },
    "./l10n": {
      "default": "./l10n/translations.json"
    }
  },
  "scripts": {
    "generate-openapi": "rm -rf src/graph/generated && docker run --rm -v \"${PWD}/src/graph:/local\" openapitools/openapi-generator-cli generate -i https://raw.githubusercontent.com/owncloud/libre-graph-api/main/api/openapi-spec/v1.0.yaml  -g typescript-axios -o /local/generated",
    "vite": "vite",
    "prepublishOnly": "rm -rf ./package && clean-publish && rm -rf package/dist/tests && find package && cat package/package.json",
    "postpublish": "rm -rf ./package"
  },
  "devDependencies": {
    "@types/luxon": "3.2.1",
    "clean-publish": "4.2.0",
    "vite-plugin-dts": "3.6.0"
  },
  "peerDependencies": {
    "@casl/ability": "^6.3.3",
    "axios": "1.6.8",
    "fast-xml-parser": "4.3.3",
    "lodash-es": "^4.17.21",
    "luxon": "3.2.1",
    "@microsoft/fetch-event-source": "^2.0.1",
    "webdav": "5.6.0",
    "xml-js": "^1.6.11",
    "zod": "3.22.4"
  }
}
