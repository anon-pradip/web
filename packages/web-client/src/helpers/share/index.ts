export * from './constants'
export * from './functions'
export * from './type'
export * from './types'
