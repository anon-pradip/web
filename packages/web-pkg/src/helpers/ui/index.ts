export * from './resourceTable'
export * from './resourceTiles'
