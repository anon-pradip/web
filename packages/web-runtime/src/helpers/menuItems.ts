import { ApplicationMenuItem } from '@ownclouders/web-pkg'

export interface MenuItem extends ApplicationMenuItem {
  id: string
  icon: string
  iconUrl: string
  title: string
  color: string
  defaultExtension: string
  active?: boolean
  path?: string
  target?: string
  url?: string
}
