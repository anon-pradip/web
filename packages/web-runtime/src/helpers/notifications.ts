import { RouteLocationNamedRaw } from 'vue-router'

export interface Notification {
  notification_id: string
  app: string
  user: string
  datetime: string
  subject: string
  message?: string
  messageRich?: string
  messageRichParameters?: any
  object_type?: string
  object_id?: string
  link?: string
  computedMessage?: string
  computedLink?: RouteLocationNamedRaw
}
