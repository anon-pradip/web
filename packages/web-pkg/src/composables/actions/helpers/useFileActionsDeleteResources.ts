import { cloneStateObject } from '../../../helpers/store'
import { isSameResource } from '../../../helpers/resource'
import { Resource, SpaceResource } from '@ownclouders/web-client'
import PQueue from 'p-queue'
import { isLocationSpacesActive } from '../../../router'
import { dirname } from 'path'
import { createFileRouteOptions } from '../../../helpers'
import { computed, unref } from 'vue'
import { queryItemAsString } from '../../appDefaults'
import { useGetMatchingSpace } from '../../spaces'
import { useRouteQuery } from '../../router'
import { useClientService } from '../../clientService'
import { useRouter } from '../../router'
import { useGettext } from 'vue3-gettext'
import { ref } from 'vue'
import {
  useMessages,
  useModals,
  useSpacesStore,
  useConfigStore,
  useSharesStore,
  useResourcesStore
} from '../../piniaStores'
import { storeToRefs } from 'pinia'
import { useDeleteWorker } from '../../webWorkers'

export const useFileActionsDeleteResources = () => {
  const configStore = useConfigStore()
  const messageStore = useMessages()
  const { showMessage, showErrorMessage } = messageStore
  const router = useRouter()
  const language = useGettext()
  const { getMatchingSpace } = useGetMatchingSpace()
  const { $gettext, $ngettext } = language
  const clientService = useClientService()
  const { dispatchModal } = useModals()
  const spacesStore = useSpacesStore()
  const sharesStore = useSharesStore()
  const { startWorker } = useDeleteWorker()

  const resourcesStore = useResourcesStore()
  const { currentFolder } = storeToRefs(resourcesStore)

  const queue = new PQueue({
    concurrency: configStore.options.concurrentRequests.resourceBatchActions
  })
  const deleteOps: Promise<void>[] = []
  const resourcesToDelete = ref([])

  const currentPageQuery = useRouteQuery('page', '1')
  const currentPage = computed(() => {
    return parseInt(queryItemAsString(unref(currentPageQuery)))
  })

  const itemsPerPageQuery = useRouteQuery('items-per-page', '1')
  const itemsPerPage = computed(() => {
    return parseInt(queryItemAsString(unref(itemsPerPageQuery)))
  })

  const resources = computed(() => {
    return cloneStateObject<Resource[]>(unref(resourcesToDelete))
  })

  const dialogTitle = computed(() => {
    const currentResources = unref(resources)
    const isFolder = currentResources[0].type === 'folder'
    let title: string = null

    if (currentResources.length === 1) {
      if (isFolder) {
        title = $gettext(
          'Permanently delete folder %{name}',
          {
            name: currentResources[0].name
          },
          true
        )
      } else {
        title = $gettext(
          'Permanently delete file %{name}',
          {
            name: currentResources[0].name
          },
          true
        )
      }
      return title
    }

    return $ngettext(
      'Permanently delete selected resource?',
      'Permanently delete %{amount} selected resources?',
      currentResources.length,
      { amount: currentResources.length.toString() },
      false
    )
  })

  const dialogMessage = computed(() => {
    const currentResources = unref(resources)
    const isFolder = currentResources[0].type === 'folder'

    if (currentResources.length === 1) {
      if (isFolder) {
        return $gettext(
          'Are you sure you want to delete this folder? All its content will be permanently removed. This action cannot be undone.'
        )
      }
      return $gettext(
        'Are you sure you want to delete this file? All its content will be permanently removed. This action cannot be undone.'
      )
    }

    return $gettext(
      'Are you sure you want to delete all selected resources? All their content will be permanently removed. This action cannot be undone.'
    )
  })

  const trashbin_deleteOp = (space: SpaceResource, resource: Resource) => {
    return clientService.webdav
      .clearTrashBin(space, {
        id: resource.id
      })
      .then(() => {
        resourcesStore.removeResources([resource])
        resourcesStore.resetSelection()

        const translated = $gettext(
          '"%{file}" was deleted successfully',
          { file: resource.name },
          true
        )
        showMessage({ title: translated })
      })
      .catch((error) => {
        if (error.statusCode === 423) {
          // TODO: we need a may retry option ....
          const p = queue.add(() => {
            return trashbin_deleteOp(space, resource)
          })
          deleteOps.push(p)
          return
        }

        console.error(error)
        const translated = $gettext('Failed to delete "%{file}"', { file: resource.name }, true)
        showErrorMessage({
          title: translated,
          errors: [error]
        })
      })
  }

  const trashbin_delete = async (space: SpaceResource) => {
    // TODO: use clear all if all files are selected
    // FIXME: Implement proper batch delete and add loading indicator
    for (const file of unref(resources)) {
      const p = queue.add(() => {
        return trashbin_deleteOp(space, file)
      })
      deleteOps.push(p)
    }

    await Promise.all(deleteOps)
  }

  const filesList_delete = (resources: Resource[]) => {
    resourcesToDelete.value = [...resources]

    const resourceSpaceMapping = unref(resources).reduce<
      Record<string, { space: SpaceResource; resources: Resource[] }>
    >((acc, resource) => {
      if (resource.storageId in acc) {
        acc[resource.storageId].resources.push(resource)
        return acc
      }

      const matchingSpace = getMatchingSpace(resource)

      if (!(matchingSpace.id in acc)) {
        acc[matchingSpace.id] = { space: matchingSpace, resources: [] }
      }

      acc[matchingSpace.id].resources.push(resource)
      return acc
    }, {})

    const originalCurrentFolderId = unref(currentFolder)?.id

    return Object.values(resourceSpaceMapping).map(
      ({ space: spaceForDeletion, resources: resourcesForDeletion }) => {
        startWorker(
          { space: spaceForDeletion, resources: resourcesForDeletion },
          async ({ successful, failed }) => {
            if (successful.length) {
              const title =
                successful.length === 1 && resources.length === 1
                  ? $gettext('"%{item}" was moved to trash bin', { item: successful[0].name })
                  : $ngettext(
                      '%{itemCount} item was moved to trash bin',
                      '%{itemCount} items were moved to trash bin',
                      successful.length,
                      { itemCount: successful.length.toString() },
                      true
                    )

              messageStore.showMessage({ title })
            }

            failed.forEach(({ status, resource }) => {
              let title = $gettext('Failed to delete "%{resource}"', { resource: resource.name })
              if (status === 423) {
                title = $gettext('Failed to delete "%{resource}" - the file is locked', {
                  resource: resource.name
                })
              }

              messageStore.showErrorMessage({ title, errors: [new Error()] })
            })

            // user hasn't navigated to another location meanwhile
            if (originalCurrentFolderId === unref(currentFolder)?.id) {
              resourcesStore.removeResources(successful)

              successful.forEach(({ id }) => {
                resourcesStore.removeSelection(id)
              })

              const activeFilesCount = resourcesStore.activeResources.length
              const pageCount = Math.ceil(unref(activeFilesCount) / unref(itemsPerPage))
              if (unref(currentPage) > 1 && unref(currentPage) > pageCount) {
                // reset pagination to avoid empty lists (happens when deleting all items on the last page)
                currentPageQuery.value = pageCount.toString()
              }
            }

            // Load quota
            if (
              isLocationSpacesActive(router, 'files-spaces-generic') &&
              !['public', 'share'].includes(spaceForDeletion?.driveType)
            ) {
              const graphClient = clientService.graphAuthenticated
              const driveResponse = await graphClient.drives.getDrive(unref(resources)[0].storageId)
              spacesStore.updateSpaceField({
                id: driveResponse.data.id,
                field: 'spaceQuota',
                value: driveResponse.data.quota
              })
            }

            if (
              unref(resourcesToDelete).length &&
              isSameResource(unref(resourcesToDelete)[0], unref(currentFolder))
            ) {
              // current folder is being deleted
              return router.push(
                createFileRouteOptions(spaceForDeletion, {
                  path: dirname(unref(resourcesToDelete)[0].path),
                  fileId: unref(resourcesToDelete)[0].parentFolderId
                })
              )
            }
          }
        )
      }
    )
  }

  const displayDialog = (space: SpaceResource, resources: Resource[]) => {
    resourcesToDelete.value = [...resources]

    dispatchModal({
      variation: 'danger',
      title: unref(dialogTitle),
      message: unref(dialogMessage),
      confirmText: $gettext('Delete'),
      onConfirm: () => trashbin_delete(space)
    })
  }

  return {
    displayDialog,
    // HACK: exported for unit tests:
    filesList_delete
  }
}
