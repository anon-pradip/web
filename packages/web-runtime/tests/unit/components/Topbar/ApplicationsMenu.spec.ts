import ApplicationsMenu from 'web-runtime/src/components/Topbar/ApplicationsMenu.vue'
import {
  RouteLocation,
  defaultComponentMocks,
  defaultPlugins,
  shallowMount
} from 'web-test-helpers'
import { mock } from 'vitest-mock-extended'
import { MenuItem } from 'web-runtime/src/helpers/menuItems'
import { SpaceResource } from '@ownclouders/web-client'

vi.mock('@ownclouders/web-pkg', async (importOriginal) => ({
  ...(await importOriginal<any>()),
  useFileActions: vi.fn(() => ({ openEditor: vi.fn() }))
}))

const menuLinks = [
  {
    icon: 'folder',
    iconUrl: undefined,
    title: 'Files',
    path: '/files'
  },
  {
    icon: 'some-icon',
    iconUrl: undefined,
    title: 'External',
    url: 'http://some.org',
    target: '_blank'
  }
] as MenuItem[]

describe('ApplicationsMenu component', () => {
  it('should render navigation with button and menu items in dropdown', () => {
    const { wrapper } = getWrapper(menuLinks)
    expect(wrapper.html()).toMatchSnapshot()
  })
})

function getWrapper(applicationsList: MenuItem[] = []) {
  const mocks = {
    ...defaultComponentMocks({
      currentRoute: mock<RouteLocation>({ query: { app: 'admin-settings' } })
    }),
    space: {
      driveType: 'personal',
      spaceRoles: { viewer: [], editor: [], manager: [] }
    } as unknown as SpaceResource
  }

  return {
    wrapper: shallowMount(ApplicationsMenu, {
      props: {
        applicationsList
      },
      global: {
        renderStubDefaultSlot: true,
        mocks,
        provide: mocks,
        plugins: [...defaultPlugins()]
      }
    })
  }
}
