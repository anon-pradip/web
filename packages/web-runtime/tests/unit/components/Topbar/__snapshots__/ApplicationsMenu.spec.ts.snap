// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ApplicationsMenu component > should render navigation with button and menu items in dropdown 1`] = `
"<nav data-v-33cec29f="" id="applications-menu" aria-label="Main navigation" class="oc-flex oc-flex-middle">
  <oc-button-stub data-v-33cec29f="" type="button" disabled="false" size="medium" arialabel="Application Switcher" submit="button" variation="brand" appearance="raw-inverse" justifycontent="center" gapsize="medium" showspinner="false" id="_appSwitcherButton" class="oc-topbar-menu-burger">
    <oc-icon-stub data-v-33cec29f="" name="grid" filltype="fill" accessiblelabel="" type="span" size="large" variation="passive" color="" class="oc-flex"></oc-icon-stub>
  </oc-button-stub>
  <oc-drop-stub data-v-33cec29f="" dropid="app-switcher-dropdown" popperoptions="[object Object]" toggle="#_appSwitcherButton" position="bottom-start" mode="click" closeonclick="true" isnested="false" paddingsize="small">
    <div data-v-33cec29f="" class="oc-display-block oc-position-relative">
      <oc-list-stub data-v-33cec29f="" raw="false" class="applications-list">
        <li data-v-33cec29f="">
          <oc-button-stub data-v-33cec29f="" type="router-link" disabled="false" size="medium" to="/files" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" showspinner="false" class="">
            <oc-application-icon-stub data-v-33cec29f="" icon="folder" colorprimary="" colorsecondary=""></oc-application-icon-stub> <span data-v-33cec29f="">Files</span>
          </oc-button-stub>
        </li>
        <li data-v-33cec29f="">
          <oc-button-stub data-v-33cec29f="" type="a" disabled="false" size="medium" href="http://some.org" target="_blank" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" showspinner="false" class="">
            <oc-application-icon-stub data-v-33cec29f="" icon="some-icon" colorprimary="" colorsecondary=""></oc-application-icon-stub> <span data-v-33cec29f="">External</span>
          </oc-button-stub>
        </li>
      </oc-list-stub>
    </div>
  </oc-drop-stub>
</nav>"
`;
